File: C#/Hash Table/Hash_Table.cs
public class Hash_Table{
    private int[] _table;
    private int _mode;
    private int _num_collisions;
    private int _num_elements;
    private int _table_size;
    private double _a_value;

    // Constructors
    public Hash_Table(int mode){
        this._mode = mode;
        this._num_collisions = 0;
        this._num_elements = 0;
        this._table_size = 1;
        this._table = new int[this._table_size];

        if(mode == 3){
            var rand = new Random();
            this._a_value = new decimal(rand.NextDouble());
        }
    }


    public Hash_Table(int mode, int table_size){
        this._mode = mode;
        this._num_collisions = 0;
        this._num_elements = 0;
        this._table_size = table_size;
        this._table = new int[this._table_size];
        
        if(mode == 3){
            var rand = new Random();
            this._a_value = new decimal(rand.NextDouble());
        }
    }


    // Private Methods
    private int num_digits(){
        int temp_size = this._table_size - 1;
        int r = 0;
        while (temp_size > 0){
            temp_size /= 10;
            r++;
        }

        return r;
    }   


    private int div_hash(int key){
        int idx = key % this._table_size;
        if(this._table[idx] != 0){
            this._num_collisions++;
            idx = this.div_hash(key + 1);
        }
        
        return idx;
    }


    private int mid_square_hash(int key){
        int idx;
        int r = this.num_digits();

        string square = (key * key).ToString();
        idx = square.Substring((square.Length / 2) - (r / 2), r).ToInt();

        if(this._table[idx] != 0){
            this._num_collisions++;
            idx = this.mid_square_hash(key + 1);
        }

        return idx;
    }


    private int digit_folding_hash(int key){
        int idx;
        int r = this.num_digits();

        string key_str = key.ToString();
        while(key_str.Length > 0){
            if(r > key_str.Length){
                r = key_str.Length;
            }

            idx += key_str.Substring(0, r).ToInt();
            key_str = key_str.Substring(r);
        }

        if(this._table[idx] != 0){
            this._num_collisions++;
            idx = this.digit_folding_hash(key + 1);
        }

        return idx;
    }


    private int mult_hash(int key, double a){
        int idx = (int)(this._table_size * ((key * a) % 1));

        if(this._table[idx] != 0){
            this._num_collisions++;
            idx = this.mult_hash(key + 1, a);
        }

        return idx;
    }


    private void resize(){
        Hash_Table new_table = new Hash_Table(this._mode, this._table_size * 2);
        for(int i = 0; i < this._table_size; i++){
            if(this._table[i] != 0){
                new_table.insert(this._table[i]);
            }
        }

        this = new_table;
    }


    // Public Methods
    public void insert(int key){
        if((double)this._num_elements / (double)this._table_size > 0.7){
            this.resize();
        }

        int idx;
        switch(this._mode){
            case 0:
                idx = this.div_hash(key);
                break;
            case 1:
                idx = this.mid_square_hash(key);
                break;
            case 2:
                idx = this.digit_folding_hash(key);
                break;
            case 3:
                idx = this.mult_hash(key, this._a_value);
                break;
            default:
                idx = this.div_hash(key);
                break;
        }

        this._table[idx] = key;
        this._num_elements++;
    }
}

File: Python/Hash Table/Hash_Table.py
import random


class HashTable:
    # Constructors
    def __init__(self, mode):
        self._mode = mode
        self._num_collisions = 0
        self._num_elements = 0
        self._table_size = 1
        self._table = [None for i in range(self._table_size)]

        if mode == 3:
            self._a_value = random.random()

    def __init__(self, mode, table_size):
        self._mode = mode
        self._num_collisions = 0
        self._num_elements = 0
        self._table_size = table_size
        self._table = [None for i in range(self._table_size)]

        if mode == 3:
            self._a_value = random.random()

    # Private Methods
    def num_digits(self):
        temp_size = self._table_size - 1
        r = 0
        while temp_size > 0:
            temp_size /= 10
            r += 1

        return r

    def div_hash(self, key):
        idx = key % self._table_size
        if self._table[idx] != 0:
            self._num_collisions += 1
            idx = self.div_hash(key + 1)

        return idx

    def mid_square_hash(self, key):
        r = self.num_digits()

        square = (key * key).ToString()
        idx = square.Substring((square.Length / 2) - (r / 2), r).ToInt()

        if self._table[idx] != 0:
            self._num_collisions += 1
            idx = self.mid_square_hash(key + 1)

        return idx

    def digit_folding_hash(self, key):
        r = self.num_digits()

        key_str = key.ToString()
        while key_str.Length > 0:
            if r > key_str.Length:
                r = key_str.Length

            idx += key_str.Substring(0, r).ToInt()
            key_str = key_str.Substring(r)

        if self._table[idx] != 0:
            self._num_collisions += 1
            idx = self.digit_folding_hash(key + 1)

        return idx

    def mult_hash(self, key, a):
        idx = (int)(self._table_size * ((key * a) % 1))

        if self._table[idx] != 0:
            self._num_collisions += 1
            idx = self.mult_hash(key + 1, a)

        return idx

    def resize(self):
        new_table = HashTable(self._mode, self._table_size * 2)
        for i in range(self._table_size):
            if self._table[i] != 0:
                new_table.insert(self._table[i])

        return new_table

    # Public Methods
    def insert(self, key):
        if self._num_elements / self._table_size > 0.7:
            self = self.resize()

        if self._mode == 0:
            idx = self.div_hash(key)
        elif self._mode == 1:
            idx = self.mid_square_hash(key)
        elif self._mode == 2:
            idx = self.digit_folding_hash(key)
        elif self._mode == 3:
            idx = self.mult_hash(key, self._a_value)
        else:
            idx = self.div_hash(key)

        self._table[idx] = key
        self._num_elements += 1
